File: internal/ast/ast.go
package ast

type Module struct {
	Pos_  Position
	Stmts []Stmt
}

func (m *Module) Pos() Position { return m.Pos_ }

type ImportItem struct {
	Name  string
	Alias string
}

type ImportStmt struct {
	Pos_  Position
	Items []ImportItem
}

func (s *ImportStmt) Pos() Position { return s.Pos_ }
func (*ImportStmt) stmtNode()       {}

type FromImportStmt struct {
	Pos_   Position
	Module string
	Items  []ImportItem
}

func (s *FromImportStmt) Pos() Position { return s.Pos_ }
func (*FromImportStmt) stmtNode()       {}

type LockStmt struct {
	Pos_  Position
	Name  string
	Value Expr
}

func (s *LockStmt) Pos() Position { return s.Pos_ }
func (*LockStmt) stmtNode()       {}

type SafeStmt struct {
	Pos_  Position
	Names []string
}

func (s *SafeStmt) Pos() Position { return s.Pos_ }
func (*SafeStmt) stmtNode()       {}

type MaskStmt struct {
	Pos_  Position
	Names []string
}

func (s *MaskStmt) Pos() Position { return s.Pos_ }
func (*MaskStmt) stmtNode()       {}

type GuardStmt struct {
	Pos_  Position
	Name  string
	Type  Expr
	Conds []Expr
}

func (s *GuardStmt) Pos() Position { return s.Pos_ }
func (*GuardStmt) stmtNode()       {}

type AssignStmt struct {
	Pos_  Position
	Left  []Expr
	Op    string
	Right Expr
}

func (s *AssignStmt) Pos() Position { return s.Pos_ }
func (*AssignStmt) stmtNode()       {}

type ExprStmt struct {
	Pos_ Position
	X    Expr
}

func (s *ExprStmt) Pos() Position { return s.Pos_ }
func (*ExprStmt) stmtNode()       {}

type Param struct {
	Name    string
	Anno    Expr
	Default Expr
	Star    bool
	DblStar bool
}

type FuncDef struct {
	Pos_       Position
	Name       string
	Params     []Param
	ReturnType Expr
	Body       []Stmt
	Decorators []Expr
}

func (s *FuncDef) Pos() Position { return s.Pos_ }
func (*FuncDef) stmtNode()       {}

type ClassDef struct {
	Pos_       Position
	Name       string
	Bases      []Expr
	Body       []Stmt
	Decorators []Expr
	Seal       bool
}

func (s *ClassDef) Pos() Position { return s.Pos_ }
func (*ClassDef) stmtNode()       {}

type OnlyStmt struct {
	Pos_    Position
	Modules []string
	Body    []Stmt
}

func (s *OnlyStmt) Pos() Position { return s.Pos_ }
func (*OnlyStmt) stmtNode()       {}

type TraceStmt struct {
	Pos_  Position
	Level string
	Args  bool
	Ret   bool
	Body  []Stmt
}

func (s *TraceStmt) Pos() Position { return s.Pos_ }
func (*TraceStmt) stmtNode()       {}

type CageStmt struct {
	Pos_      Position
	HasTime   bool
	MaxTime   float64
	HasMem    bool
	MaxMemory int64
	Body      []Stmt
}

func (s *CageStmt) Pos() Position { return s.Pos_ }
func (*CageStmt) stmtNode()       {}

type ElifClause struct {
	Pos_ Position
	Cond Expr
	Body []Stmt
}

type IfStmt struct {
	Pos_  Position
	Cond  Expr
	Then  []Stmt
	Elifs []ElifClause
	Else  []Stmt
}

func (s *IfStmt) Pos() Position { return s.Pos_ }
func (*IfStmt) stmtNode()       {}

type ForStmt struct {
	Pos_   Position
	Target Expr
	Iter   Expr
	Body   []Stmt
	Else   []Stmt
}

func (s *ForStmt) Pos() Position { return s.Pos_ }
func (*ForStmt) stmtNode()       {}

type WhileStmt struct {
	Pos_ Position
	Cond Expr
	Body []Stmt
	Else []Stmt
}

func (s *WhileStmt) Pos() Position { return s.Pos_ }
func (*WhileStmt) stmtNode()       {}

type ReturnStmt struct {
	Pos_  Position
	Value Expr
}

func (s *ReturnStmt) Pos() Position { return s.Pos_ }
func (*ReturnStmt) stmtNode()       {}

type RaiseStmt struct {
	Pos_ Position
	Exc  Expr
	From Expr
}

func (s *RaiseStmt) Pos() Position { return s.Pos_ }
func (*RaiseStmt) stmtNode()       {}

type ExceptClause struct {
	Pos_ Position
	Type Expr
	Name string
	Body []Stmt
}

type TryStmt struct {
	Pos_     Position
	Body     []Stmt
	Handlers []ExceptClause
	Else     []Stmt
	Finally  []Stmt
}

func (s *TryStmt) Pos() Position { return s.Pos_ }
func (*TryStmt) stmtNode()       {}

type PassStmt struct {
	Pos_ Position
}

func (s *PassStmt) Pos() Position { return s.Pos_ }
func (*PassStmt) stmtNode()       {}

type BreakStmt struct {
	Pos_ Position
}

func (s *BreakStmt) Pos() Position { return s.Pos_ }
func (*BreakStmt) stmtNode()       {}

type ContinueStmt struct {
	Pos_ Position
}

func (s *ContinueStmt) Pos() Position { return s.Pos_ }
func (*ContinueStmt) stmtNode()       {}

type DeleteStmt struct {
	Pos_    Position
	Targets []Expr
}

func (s *DeleteStmt) Pos() Position { return s.Pos_ }
func (*DeleteStmt) stmtNode()       {}

type Name struct {
	Pos_ Position
	Name string
}

func (e *Name) Pos() Position { return e.Pos_ }
func (*Name) exprNode()       {}

type IntLit struct {
	Pos_  Position
	Value string
}

func (e *IntLit) Pos() Position { return e.Pos_ }
func (*IntLit) exprNode()       {}

type FloatLit struct {
	Pos_  Position
	Value string
}

func (e *FloatLit) Pos() Position { return e.Pos_ }
func (*FloatLit) exprNode()       {}

type StringLit struct {
	Pos_  Position
	Value string
}

func (e *StringLit) Pos() Position { return e.Pos_ }
func (*StringLit) exprNode()       {}

type EllipsisLit struct {
	Pos_ Position
}

func (e *EllipsisLit) Pos() Position { return e.Pos_ }
func (*EllipsisLit) exprNode()       {}

type ListLit struct {
	Pos_  Position
	Elems []Expr
}

func (e *ListLit) Pos() Position { return e.Pos_ }
func (*ListLit) exprNode()       {}

type TupleLit struct {
	Pos_  Position
	Elems []Expr
	Paren bool
}

func (e *TupleLit) Pos() Position { return e.Pos_ }
func (*TupleLit) exprNode()       {}

type DictLit struct {
	Pos_ Position
	Keys []Expr
	Vals []Expr
}

func (e *DictLit) Pos() Position { return e.Pos_ }
func (*DictLit) exprNode()       {}

type SetLit struct {
	Pos_  Position
	Elems []Expr
}

func (e *SetLit) Pos() Position { return e.Pos_ }
func (*SetLit) exprNode()       {}

type KwArg struct {
	Pos_  Position
	Name  string
	Value Expr
}

type CallExpr struct {
	Pos_    Position
	Func    Expr
	Args    []Expr
	Kwargs  []KwArg
	Star    Expr
	DblStar Expr
}

func (e *CallExpr) Pos() Position { return e.Pos_ }
func (*CallExpr) exprNode()       {}

type AttrExpr struct {
	Pos_ Position
	X    Expr
	Name string
}

func (e *AttrExpr) Pos() Position { return e.Pos_ }
func (*AttrExpr) exprNode()       {}

type SliceExpr struct {
	Pos_ Position
	Lo   Expr
	Hi   Expr
	Step Expr
}

func (e *SliceExpr) Pos() Position { return e.Pos_ }
func (*SliceExpr) exprNode()       {}

type SubscriptExpr struct {
	Pos_  Position
	X     Expr
	Index Expr
}

func (e *SubscriptExpr) Pos() Position { return e.Pos_ }
func (*SubscriptExpr) exprNode()       {}

type BinOpExpr struct {
	Pos_ Position
	Op   string
	X    Expr
	Y    Expr
}

func (e *BinOpExpr) Pos() Position { return e.Pos_ }
func (*BinOpExpr) exprNode()       {}

type UnaryOpExpr struct {
	Pos_ Position
	Op   string
	X    Expr
}

func (e *UnaryOpExpr) Pos() Position { return e.Pos_ }
func (*UnaryOpExpr) exprNode()       {}

type BoolOpExpr struct {
	Pos_ Position
	Op   string
	X    Expr
	Y    Expr
}

func (e *BoolOpExpr) Pos() Position { return e.Pos_ }
func (*BoolOpExpr) exprNode()       {}

type CompareExpr struct {
	Pos_ Position
	X    Expr
	Ops  []string
	Ys   []Expr
}

func (e *CompareExpr) Pos() Position { return e.Pos_ }
func (*CompareExpr) exprNode()       {}

type CondExpr struct {
	Pos_ Position
	Cond Expr
	Then Expr
	Else Expr
}

func (e *CondExpr) Pos() Position { return e.Pos_ }
func (*CondExpr) exprNode()       {}

type CompClause struct {
	Target Expr
	Iter   Expr
	Ifs    []Expr
}

type ListComp struct {
	Pos_    Position
	Elem    Expr
	Clauses []CompClause
}

func (e *ListComp) Pos() Position { return e.Pos_ }
func (*ListComp) exprNode()       {}

File: src/ast.rs
// ast.rs：Rust 版 AST 节点定义（与 Go 版 internal/ast 对应，仅 check 管线使用）。
use crate::diagnostic::Position;

#[derive(Debug, Clone, PartialEq)]
pub struct Module {
    pub pos: Position,
    pub stmts: Vec<Stmt>,
}

impl Module {
    pub fn new(pos: Position, stmts: Vec<Stmt>) -> Self {
        Module { pos, stmts }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ImportItem {
    pub name: String,
    pub alias: String,
}

impl ImportItem {
    // Module 返回导入的顶层模块名（import os.path → "os"）。
    pub fn module(&self) -> &str {
        match self.name.find('.') {
            Some(i) => &self.name[..i],
            None => &self.name,
        }
    }

    // TopName 返回绑定名（import os.path → "os"；import os.path as p → "p"）。
    pub fn top_name(&self) -> &str {
        let n = if self.alias.is_empty() {
            &self.name
        } else {
            &self.alias
        };
        match n.find('.') {
            Some(i) => &n[..i],
            None => n,
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub name: String,
    pub anno: Option<Box<Expr>>,
    pub default: Option<Box<Expr>>,
    pub star: bool,
    pub dbl_star: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ElifClause {
    pub pos: Position,
    pub cond: Box<Expr>,
    pub body: Vec<Stmt>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExceptClause {
    pub pos: Position,
    pub ty: Option<Box<Expr>>,
    pub name: String,
    pub body: Vec<Stmt>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct KwArg {
    pub pos: Position,
    pub name: String,
    pub value: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct CompClause {
    pub target: Box<Expr>,
    pub iter: Box<Expr>,
    pub ifs: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    Import {
        pos: Position,
        items: Vec<ImportItem>,
    },
    FromImport {
        pos: Position,
        module: String,
        items: Vec<ImportItem>,
    },
    Lock {
        pos: Position,
        name: String,
        value: Option<Box<Expr>>,
    },
    Safe {
        pos: Position,
        names: Vec<String>,
    },
    Mask {
        pos: Position,
        names: Vec<String>,
    },
    Guard {
        pos: Position,
        name: String,
        ty: Option<Box<Expr>>,
        conds: Vec<Expr>,
    },
    Assign {
        pos: Position,
        left: Vec<Expr>,
        op: String,
        right: Box<Expr>,
    },
    ExprStmt {
        pos: Position,
        x: Box<Expr>,
    },
    FuncDef {
        pos: Position,
        name: String,
        params: Vec<Param>,
        return_type: Option<Box<Expr>>,
        body: Vec<Stmt>,
        decorators: Vec<Expr>,
    },
    ClassDef {
        pos: Position,
        name: String,
        bases: Vec<Expr>,
        body: Vec<Stmt>,
        decorators: Vec<Expr>,
        seal: bool,
    },
    Only {
        pos: Position,
        modules: Vec<String>,
        body: Vec<Stmt>,
    },
    Trace {
        pos: Position,
        level: String,
        args: bool,
        ret: bool,
        body: Vec<Stmt>,
    },
    Cage {
        pos: Position,
        has_time: bool,
        max_time: f64,
        has_mem: bool,
        max_memory: i64,
        body: Vec<Stmt>,
    },
    If {
        pos: Position,
        cond: Box<Expr>,
        then: Vec<Stmt>,
        elifs: Vec<ElifClause>,
        els: Vec<Stmt>,
    },
    For {
        pos: Position,
        target: Box<Expr>,
        iter: Box<Expr>,
        body: Vec<Stmt>,
        els: Vec<Stmt>,
    },
    While {
        pos: Position,
        cond: Box<Expr>,
        body: Vec<Stmt>,
        els: Vec<Stmt>,
    },
    Return {
        pos: Position,
        value: Option<Box<Expr>>,
    },
    Raise {
        pos: Position,
        exc: Option<Box<Expr>>,
        from: Option<Box<Expr>>,
    },
    Try {
        pos: Position,
        body: Vec<Stmt>,
        handlers: Vec<ExceptClause>,
        els: Vec<Stmt>,
        finally: Vec<Stmt>,
    },
    Pass {
        pos: Position,
    },
    Break {
        pos: Position,
    },
    Continue {
        pos: Position,
    },
    Delete {
        pos: Position,
        targets: Vec<Expr>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Name {
        pos: Position,
        name: String,
    },
    IntLit {
        pos: Position,
        value: String,
    },
    FloatLit {
        pos: Position,
        value: String,
    },
    StringLit {
        pos: Position,
        value: String,
    },
    EllipsisLit {
        pos: Position,
    },
    ListLit {
        pos: Position,
        elems: Vec<Expr>,
    },
    TupleLit {
        pos: Position,
        elems: Vec<Expr>,
        paren: bool,
    },
    DictLit {
        pos: Position,
        keys: Vec<Expr>,
        vals: Vec<Expr>,
    },
    SetLit {
        pos: Position,
        elems: Vec<Expr>,
    },
    Call {
        pos: Position,
        func: Box<Expr>,
        args: Vec<Expr>,
        kwargs: Vec<KwArg>,
        star: Option<Box<Expr>>,
        dbl_star: Option<Box<Expr>>,
    },
    Attr {
        pos: Position,
        x: Box<Expr>,
        name: String,
    },
    Slice {
        pos: Position,
        lo: Option<Box<Expr>>,
        hi: Option<Box<Expr>>,
        step: Option<Box<Expr>>,
    },
    Subscript {
        pos: Position,
        x: Box<Expr>,
        index: Box<Expr>,
    },
    BinOp {
        pos: Position,
        op: String,
        x: Box<Expr>,
        y: Box<Expr>,
    },
    UnaryOp {
        pos: Position,
        op: String,
        x: Box<Expr>,
    },
    BoolOp {
        pos: Position,
        op: String,
        x: Box<Expr>,
        y: Box<Expr>,
    },
    Compare {
        pos: Position,
        x: Box<Expr>,
        ops: Vec<String>,
        ys: Vec<Expr>,
    },
    Cond {
        pos: Position,
        cond: Box<Expr>,
        then: Box<Expr>,
        els: Box<Expr>,
    },
    ListComp {
        pos: Position,
        elem: Box<Expr>,
        clauses: Vec<CompClause>,
    },
}

impl Expr {
    pub fn pos(&self) -> Position {
        match self {
            Expr::Name { pos, .. }
            | Expr::IntLit { pos, .. }
            | Expr::FloatLit { pos, .. }
            | Expr::StringLit { pos, .. }
            | Expr::EllipsisLit { pos }
            | Expr::ListLit { pos, .. }
            | Expr::TupleLit { pos, .. }
            | Expr::DictLit { pos, .. }
            | Expr::SetLit { pos, .. }
            | Expr::Call { pos, .. }
            | Expr::Attr { pos, .. }
            | Expr::Slice { pos, .. }
            | Expr::Subscript { pos, .. }
            | Expr::BinOp { pos, .. }
            | Expr::UnaryOp { pos, .. }
            | Expr::BoolOp { pos, .. }
            | Expr::Compare { pos, .. }
            | Expr::Cond { pos, .. }
            | Expr::ListComp { pos, .. } => *pos,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn import_item_module() {
        let it = ImportItem {
            name: "os.path".into(),
            alias: String::new(),
        };
        assert_eq!(it.module(), "os");
        assert_eq!(it.top_name(), "os");
        let it = ImportItem {
            name: "os.path".into(),
            alias: "p".into(),
        };
        assert_eq!(it.top_name(), "p");
        let it = ImportItem {
            name: "os".into(),
            alias: String::new(),
        };
        assert_eq!(it.module(), "os");
        assert_eq!(it.top_name(), "os");
    }
}
